BDArmory/GameModes/Waypoints/CourseBuilderGUI.cs

```diff
@@ -0,0 +1,679 @@
+                AddGate(newName);
+                startTime = Time.time;
+            }
+        }
+
+        void SnapCameraToGate() //this is locking cam to gate, can zoom, but can't rotate cam, can't return to vessel, cam sometimes shoots into orbit; Ask doc about Camera modification - FIXME
+        {
+            if (selected_gate_index < 0 || selected_index < 0 || selected_index >= WaypointCourses.CourseLocations.Count || selected_gate_index >= WaypointCourses.CourseLocations[selected_index].waypoints.Count) return; // FIXME What transform holds the spawnPoint?
+            // return; //returning until I can get Doc to look at this and tell me what I'm doing wrong
+            //ideally, camera should snap to selected gate, but still allow camera movement to allow user to look around, return cam to vessel when coursebuilder GUi closed.
+            var flightCamera = FlightCamera.fetch;
+            var cameraHeading = FlightCamera.CamHdg;
+            var cameraPitch = FlightCamera.CamPitch;
+            var distance = 1000;
+            //if (FlightGlobals.ActiveVessel != null && FlightGlobals.ActiveVessel.PatchedConicsAttached) FlightGlobals.ActiveVessel.DetachPatchedConicsSolver();
+            Waypoint gate = WaypointCourses.CourseLocations[selected_index].waypoints[selected_gate_index];
+            var terrainAltitude = FlightGlobals.currentMainBody.TerrainAltitude(gate.location.x, gate.location.y);
+            var spawnPoint = FlightGlobals.currentMainBody.GetWorldSurfacePosition(gate.location.x, gate.location.y, terrainAltitude + gate.location.z);
+            // FloatingOrigin.SetOffset(gate.location); // This adjusts local coordinates, such that spawnPoint is (0,0,0).
+            // var radialUnitVector = -FlightGlobals.currentMainBody.transform.position.normalized;
+            // var cameraPosition = Vector3.RotateTowards(distance * radialUnitVector, Quaternion.AngleAxis(cameraHeading * Mathf.Rad2Deg, radialUnitVector) * -VectorUtils.GetNorthVector(spawnPoint, FlightGlobals.currentMainBody), 70f * Mathf.Deg2Rad, 0);
+
+            // flightCamera.transform.parent = loadedGates[selected_gate_index].transform;
+            if (moddingSpawnPoint)
+            {
+                float terrainAlt = (float)FlightGlobals.currentMainBody.TerrainAltitude(WaypointCourses.CourseLocations[selected_index].spawnPoint.x, WaypointCourses.CourseLocations[selected_index].spawnPoint.y);
+                Vector3d SpawnCoords = new Vector3((float)WaypointCourses.CourseLocations[selected_index].spawnPoint.x, (float)WaypointCourses.CourseLocations[selected_index].spawnPoint.y, BDArmorySettings.VESSEL_SPAWN_ALTITUDE + terrainAlt);
+                flightCamera.transform.parent.position = SpawnCoords;
+            }
+            else
+                flightCamera.transform.parent.position = loadedGates[selected_gate_index].transform.position; //working, but has the pivot point near-ish gate, not gate itself
```

Comment:
The order of setting the transform position and setting the floating origin offset might be better in the other order (might need to play with it a bit to see what works best, I haven't actually tested this). Something like:
```csharp
Vector3 origin;
if (moddingSpawnPoint)
{
    ...
    origin = SpawnCoords;
}
else
{
   origin = loadedGates[selected_gate_index].transform.position;
}
FloatingOrigin.SetOffset(origin);
flightCamera.transform.parent.position = Vector3.zero;
```

Reply:
This works.